### PR TITLE
Feature/new chat

### PR DESCRIPTION
**Pull request by connor4312**
Depends on the previous #32. Adds a "tag" component that detects valid user tags like @connor4312, normalizing casing and inserting their ID. The new frontend will be able to use this to automatically link to suers.

### CODE REVIEW

**Comment by connor4312** [comment]
Also **important:** this is a breaking change unfortunately, so it should not be deployed until I have the refactored frontend chat ready. This is the only API-side change needed for the refactors, however, so I'm submitting it for review now 😄

**Comment by SReject** [comment]
The `tag` component should also include a property containing the original client-sent text:

If the client sent `@sreject` the tag should be akin to:
```json
    "type": "tag",
    "username": "SReject",
    "id": 89751,
    "text": "@sreject"
```

**Comment by connor4312** [comment]
Added in b410840

File: user-tags/chat/tags.js [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+                continue;
+            }
+
+            const username = text.slice(prefix.length).toLowerCase();
+            const cached = this._getCachedData(username);
+
+            if (cached) {
+                message[i] = _.assign({ text }, cached);
+            } else {
+                todo.push({ i, text, username: username });
```
Comment (UnwrittenFun) [inline user-tags/chat/tags.js]: Can use shorthand for username

**Comment by MattJustMatt** [comment]
Waiting on merge for this one on frontend changes

**Comment by MattJustMatt** [comment]
@connor4312 Need to get @jackcook and @vijaysharm on board to get apps in-sync here. Any way we could add in some backwards compatibility?

**Comment by connor4312** [comment]
In general the backwards-compatible way to do anything involving chat moving forward is to just use the `text` property of an object if you don't recognize it. Unfortunately it is rather difficult to add backwards compatibility here outside of that. Is there time to get an update out to the apps some time next week?

**Comment by MattJustMatt** [comment]
Yep, the big refactor is rolling out next week (will likely be approved on Wednesday) so we can coordinate around that.

**Comment by MattJustMatt** [comment]
I'm worried about older versions of the app crashing because of this though. We should probably run this through some staging tests to see how the current versions of the app handle it. @connor4312 @jackcook @vijaysharm

**Comment by vijaysharm** [comment]
wait, breaking change? so android and ios apps that are out there now will be broken?

**Comment by MattJustMatt** [comment]
Not yet but if this were merged. mv -> Slack (this is a public repo) @vijaysharm @jackcook @connor4312

**Comment by connor4312** [comment]
Maybe, it depends how they currently handle chat messages. If they die when they see a message segment they don't recognize, then it will cause them to break.

**Comment by ttaylorr** [comment]
LGTM

**Pull request merged**